### PR TITLE
Allow ingress without host rule

### LINKED ISSUE #11770: [stable/jenkins] Allow ingress without specifying host
I'm not sure if this is a feature request or a misunderstanding (can reword as a proper template feature request once I clear that up), but considering you can create an ingress in Kube without specifying a host, why does the chart force the setting of Master.HostName before even generating an ingress yaml ? I don't want a host entry in my ingress and removing it afterwards is obviously less than ideal. Am I missing a trick or would this be a nice enhancement?

### PR DESCRIPTION
**Pull request by admoriarty**
#### What this PR does / why we need it:
Addresses the desire to create an ingress without a host rule

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #11770

#### Special notes for your reviewer:
My version change may be wrong considering this probably is a breaking change...perhaps the top level if statement in the ingress template could tolerate the new boolean OR the HostPath being set for backwards compatibility?

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x ] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by maorfr** [comment]
Follow the DCO Details link to see how to solve it

File: stable/jenkins/values.yaml
Comment (maorfr) [inline stable/jenkins/values.yaml]: I would expect this section to be more similar to:
https://github.com/helm/charts/blob/master/stable/grafana/values.yaml#L83

File: stable/jenkins/values.yaml
Comment (maorfr) [inline stable/jenkins/values.yaml]: I think that while we are visiting this area, maybe we can remove this option from values and hard code it in the resource?

File: stable/jenkins/templates/jenkins-master-ingress.yaml
Comment (maorfr) [inline stable/jenkins/templates/jenkins-master-ingress.yaml]: maybe we can move the `hostname` field to the end of the section to avoid duplicating `http`?

**Review by maorfr — CHANGES_REQUESTED** [review]

**Comment by admoriarty** [comment]
> Follow the DCO Details link to see how to solve it

Thanks, was just being lazy and trying to keep all edits online, sorted now ;)

File: stable/jenkins/values.yaml
Comment (admoriarty) [reply stable/jenkins/values.yaml]: Sure, can be, I was just scanning other boolean usages in the same chart, but I see that "Enabled" within the immediate context is used too. I'll make the change...the only weird thing I thought was that HostName was left outside the ingress context but controlled it before which led me down that same path?

Comment (maorfr) [reply stable/jenkins/values.yaml]: I completely understand why you chose this path.
Nevertheless, this chart is currently not following the [chart best practices](https://helm.sh/docs/chart_best_practices).

If you are willing to extend your contribution and make the entire `ingress` section more compliant with the above mentioned guidelines, it will be highly appreciated!

You can look at the same section in the [grafana](https://github.com/helm/charts/blob/master/stable/grafana/values.yaml#L82) chart for reference ;)

File: stable/jenkins/values.yaml
Comment (admoriarty) [reply stable/jenkins/values.yaml]: The ApiVersion? Happy to move it once you confirm.

Comment (maorfr) [reply stable/jenkins/values.yaml]: exactly right.

File: stable/jenkins/templates/jenkins-master-ingress.yaml
Comment (admoriarty) [reply stable/jenkins/templates/jenkins-master-ingress.yaml]: Ha, makes a lot more sense than what I did :)

File: stable/jenkins/values.yaml
Comment (admoriarty) [reply stable/jenkins/values.yaml]: Already updated, please check

File: stable/jenkins/values.yaml
Comment (admoriarty) [reply stable/jenkins/values.yaml]: Updated

File: stable/jenkins/README.md
Comment (maorfr) [inline stable/jenkins/README.md]: I would even go as far as moving the HostName field into ingress.
What do you think?

File: stable/jenkins/README.md
Comment (maorfr) [inline stable/jenkins/README.md]: This is starting to look so much better!
Can we change the entire ingress section to follow the [naming best practices](https://helm.sh/docs/chart_best_practices/#naming-conventions)?

File: stable/jenkins/Chart.yaml
Comment (maorfr) [inline stable/jenkins/Chart.yaml]: Lets do a minor version bump

File: stable/jenkins/README.md
Comment (admoriarty) [reply stable/jenkins/README.md]: As an outsider I would say it makes perfect sense, but I've no context as to why it lived outside that block in the first place :) Perfectly happy to move it though...at what point does this affect the versioning changes btw (this is definitely a breaking change, no?)

File: stable/jenkins/Chart.yaml
Comment (admoriarty) [reply stable/jenkins/Chart.yaml]: ok

File: stable/jenkins/README.md
Comment (maorfr) [reply stable/jenkins/README.md]: you are right. this is why i suggested to bump the minor version instead of the patch.

Comment (maorfr) [reply stable/jenkins/README.md]: and as far as breaking changes go, i think that we are good to go as long as everything is well documented.

File: stable/jenkins/README.md
Comment (admoriarty) [reply stable/jenkins/README.md]: lowercase and camelcase? Wouldn't that be completely inconsistent with the rest of this chart then?

Comment (maorfr) [reply stable/jenkins/README.md]: it will, but my goal is to have the entire chart follow the best practices. it is a work in progress :)

Comment (admoriarty) [reply stable/jenkins/README.md]: So something like;

Comment (admoriarty) [reply stable/jenkins/README.md]: So something like;
`Master.ingress.enabled`  
`Master.ingress.hostName` 
`Master.ingress.annotations` 
`Master.ingress.labels`    
`Master.ingress.path` 
`Master.ingress.TLS`

Not sure about TLS versus tls versus tLS :D

Comment (maorfr) [reply stable/jenkins/README.md]: Yes, just like that! TLS is the way to go :)

Comment (maorfr) [reply stable/jenkins/README.md]: let me correct myself, `tls` is the way to go :)

Comment (admoriarty) [reply stable/jenkins/README.md]: done

File: stable/jenkins/values.yaml
Comment (maorfr) [inline stable/jenkins/values.yaml]: path

**Comment by maorfr** [comment]
/assign
I think that there are some changes to be done in config.yaml
And a new PR was just merged, so you would need to resolve the conflict in Chart.yaml

Thanks!

**Comment by admoriarty** [comment]
> /assign
> I think that there are some changes to be done in config.yaml
> And a new PR was just merged, so you would need to resolve the conflict in Chart.yaml
> 
> Thanks!

Ah, so now I see why HostName was outside the Ingress block...it has an independent purpose. You sure it still makes sense to move it within the ingress context? I'll make the config.yaml changes accordingly once confirmed.

**Comment by maorfr** [comment]
I think it still makes sense, I saw this before we started this PR.

Do you think we should add a note to NOTES to let users know that this chart is undergoing breaking changes?

**Comment by admoriarty** [comment]
I'll make the config.yaml changes to keep in line with the rest so, as for NOTES I have no opinion on managing this repo so I'll let you or another make that call :D

**Comment by maorfr** [comment]
I will be happy to see such a note. Will you be willing to add it?

File: stable/jenkins/values.yaml
Comment (maorfr) [reply stable/jenkins/values.yaml]: path instead of Path

**Comment by admoriarty** [comment]
I can add anything you want but it might be best for an owner to dictate the contents and positioning within that file ;) Also, and again this is just coming from an outsider with little context to managing these charts, why would we be talking about notes around breaking changes but not updating the versioning to reflect a breaking change with a major version bump (again I'm assuming a style of versioning perhaps from my own experiences!)

**Comment by maorfr** [comment]
We are still within major version 0, so lets stick to a minor bump. And don't worry about the note, I'll take care of it at a later time.

Appeeciate your cooperation :)

File: stable/jenkins/README.md
Comment (maorfr) [inline stable/jenkins/README.md]: enabled instead of Enabled

File: stable/jenkins/README.md
Comment (maorfr) [inline stable/jenkins/README.md]: Ingress host name instead of path

File: stable/jenkins/values.yaml
Comment (maorfr) [inline stable/jenkins/values.yaml]: path instead of Path (even though that is commented out :) )

**Comment by maorfr** [comment]
/ok-to-test

This is an awesome set of changes.
If we can get the last few small commets sorted, we'll get this merged! :)

File: stable/jenkins/values.yaml
Comment (admoriarty) [reply stable/jenkins/values.yaml]: updated

File: stable/jenkins/README.md
Comment (admoriarty) [reply stable/jenkins/README.md]: updated

File: stable/jenkins/README.md
Comment (admoriarty) [reply stable/jenkins/README.md]: fixed

Comment (maorfr) [reply stable/jenkins/README.md]: did you push these changes? i can't see them. same for the other 2 changes

**Comment by maorfr** [comment]
/approve
/lgtm

Thanks for your contribution! Looking forward to seeing more from you in the future ;)

**Comment by jeff-knurek** [comment]
@maorfr 
> We are still within major version 0, so lets stick to a minor bump. And don't worry about the note, I'll take care of it at a later time.

I'm too late to this being merged, but I think that changing `Master.Ingress` to `Master.ingress` (and removing `Master.HostName`) is a pretty significant breaking change. The structure of the values changing should have been a major version bump. 

Going from 0 -> 1 is just a normal change in major version and isn't life changing. _Major_ changes require major version updates.

**Comment by maorfr** [comment]
I both agree and disagree at the same time :)

going from n -> n+1 is normal. going from 0 -> 1 a declaration.
just for the record - https://semver.org/#spec-item-4
this chart is one of the first, and as such, was created before best practices were much of a thing. the major version 0 is a reminder for that.

on the other hand - this chart is a part of the "stable" repo, so this is definitely confusing.

moving forward - we are going for a chart version 1.0 that will follow best practices as we know them. this will probably include many breaking changes. we have thought about it, but could not reach a solution that will preserve support for the old values while creating new ones. if you have thoughts about this - please let me know!

in any case, i apologize if this has caused any problems on your end. we have tried to be as explicit as we can through the NOTES.

again, if you have any thoughts on how we can make this transition better - let me know!
i can also be reached on k8s slack if you prefer!

**Comment by mrdunski** [comment]
Would be great to not break backward compatibility. 
After update with default values, helm removed ingress and tls configuration.
So, update is not so straight forward. Could you revert that change?

Or at least prepare upgrade instruction for those who use ingress? 
Also I would suggest to bump major version number for changes that breaks compatibility.

**Comment by maorfr** [comment]
@mrdunski please see #12935.

Instructions are present in NOTES.txt.

**Comment by mrdunski** [comment]
@maorfr 

```
#### This chart is undergoing work which may result in breaking changes. ####
#### Please consult the following table and the README ####

Master.HostName --> Master.ingress.hostName
New value - Master.ingress.enabled
Master.Ingress.ApiVersion --> Master.ingress.apiVersion
Master.Ingress.Annotations --> Master.ingress.annotations
Master.Ingress.Labels --> Master.ingress.labels
Master.Ingress.Path --> Master.ingress.path
Master.Ingress.TLS --> Master.ingress.tls
Master.OwnSshKey, a bool, has been replaced with Master.AdminSshKey, which is expected to be a string containing the actual key
```
Do you mean this? I've noticed them but it seems to be a bit too laconic, so I haven't realize those are migration instructions. Also NOTES.txt meant to be shown after you do the upgrade.
IMO it would be better to move it to the README.

**Comment by maorfr** [comment]
as stated in #12935, we are going to create a copy of the chart in incubator and work on it there (pending maintainers approval). sorry for any inconvenience that may have been caused!

**Pull request merged**
